core/domain/html_validation_service.py

```diff
@@ -0,0 +1,767 @@
+        RTE_TYPE]['ALLOWED_PARENT_LIST']
+    allowed_tag_list = feconf.RTE_CONTENT_SPEC[RTE_TYPE]['ALLOWED_TAG_LIST']
+
+    is_invalid = False
+
+    # Text with no parent tag is also invalid.
+    for content in soup.contents:
+        if not content.name:
+            is_invalid = True
+
+    for tag in soup.findAll():
+        # Checking for tags not allowed in RTE.
+        if tag.name not in allowed_tag_list:
+            if 'invalidTags' in err_dict:
+                err_dict['invalidTags'].append(tag.name)
+            else:
+                err_dict['invalidTags'] = [tag.name]
+            is_invalid = True
+        # Checking for parent-child relation that are not
+        # allowed in RTE.
+        parent = tag.parent.name
+        if (tag.name in allowed_tag_list) and (
+                parent not in allowed_parent_list[tag.name]):
+            if tag.name in err_dict:
+                err_dict[tag.name].append(parent)
+            else:
+                err_dict[tag.name] = [parent]
+            is_invalid = True
+
+    return is_invalid
+
+
+def add_caption_attr_to_image(html_string):
+    """Adds caption attribute to all oppia-noninteractive-image tags.
+
+    Args:
+        html_string. str: HTML string in which the caption attribute is to be
+            added.
+
+    Returns:
+        str. Updated HTML string with the caption attribute for all
+            oppia-noninteractive-image tags.
+    """
+    soup = bs4.BeautifulSoup(
+        html_string.encode(encoding='utf-8'), 'html.parser')
+
+    for image in soup.findAll(name='oppia-noninteractive-image'):
+        attrs = image.attrs
+        if 'caption-with-value' not in attrs:
+            image['caption-with-value'] = escape_html(json.dumps(''))
+
+    return unicode(soup)
+
+
+def validate_customization_args(html_list):
+    """Validates customization arguments of Rich Text Components in a list of
+    html string.
+
+    Args:
+        html_list: list(str). List of html strings to be validated.
+
+    Returns:
+        dict: Dictionary of all the invalid customisation args where
+            key is a Rich Text Component and value is the invalid html string.
+    """
+    # Dictionary to hold html strings in which customization arguments
+    # are invalid.
+    err_dict = {}
+    RICH_TEXT_COMPONENTS = OPPIA_INLINE_COMPONENTS + OPPIA_BLOCK_COMPONENTS
+
+    tag_dict = {}
+    for html_string in html_list:
+        soup = bs4.BeautifulSoup(
+            html_string.encode(encoding='utf-8'), 'html.parser')
+
+        for tag_name in RICH_TEXT_COMPONENTS:
+            for tag in soup.findAll(name=tag_name):
+                tag_dict[tag] = html_string
+
+    for tag in tag_dict:
+        html_string = tag_dict[tag]
+        err_msg_list = list(_validate_customization_args_in_tag(tag))
+        for err_msg in err_msg_list:
+            if err_msg:
+                if err_msg not in err_dict:
+                    err_dict[err_msg] = [html_string]
```

Comment:
I wanted to store the complete html string since I wanted to make sure that when we do the migration we have the complete string in which the tag is present instead of just the tag to avoid any mistakes.